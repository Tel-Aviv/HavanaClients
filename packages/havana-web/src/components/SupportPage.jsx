import React, { useState, useEffect, useReducer, useContext } from 'react';
import { useParams } from 'react-router-dom';
import { Collapse, Button, 
    Layout, Row, Col, Typography } from 'antd';
import { CheckCircleOutlined } from '@ant-design/icons';
const { Panel } = Collapse;
const { Title } = Typography;

import { useTranslation } from "react-i18next";

import { API, DEFAULT_BASE_URL } from '../utils';
import { DataContext } from '../DataContext';

const reducer = (state, action) => {
    if( action.type == 'UPDATE' ) {

        return state.map( step => {
            if( step.url === action.payload.url ) {
                return action.payload;
            } else 
                return {...step}
        }) 
    }
}

class Step {

    constructor(name, method, url, params, postProcess) {
        this.name = name;
        this.method = method;
        this.url = url;
        this.params = params;
        this.postProcess = postProcess;
        this.queryString = '';
    }
    
    getQueryString() { // just serialize params

        let _queryString = '';
        if( this.params ) {
            let firstParam = '?'
            Object.entries(this.params).forEach(
                ([key, value]) => {
                    _queryString += `${firstParam}${key}=${value}`;
                    firstParam = '&';
                }
            );
        }

        return _queryString;
    }

    async process() {
        
        try {
            // const resp = await API.get(this.url, {
            //     params: this.params
            // })
            const resp = await API({
                method: this.method,
                url: this.url,
                params: this.params
            })
            this.status = resp.status;
            this.result = resp.data;
            if( this.postProcess ) {
                this.postProcess(resp);
            }
            this.queryString = this.getQueryString();

        } catch(error) {

            let _message = `${error.message} when reaching ${DEFAULT_BASE_URL}/${this.url}`;
            let _status = 0;
            const { response } = error;
            if( response ) {
                _message = `${response.config.url} => ${response.statusText} (${response.status})`;
                _status = response.status;
            }

            this.status = _status || 0;
            this.result = _message;
        }   
    }
}

const SupportPage = () => {

    // The order of following hooks is important here.
    // Firstly useParams() and useContext to populate steps array.
    // Later this array will be passed to useReducer() as initial state.
    const routeParams = useParams();
    const context = useContext(DataContext);
    const [activePanelKeys, setActivePanelKeys] = useState([]);
    const [expanded, setExpanded] = useState(false);
    const [loading , setLoading] = useState(false);

    const { t } = useTranslation();

    // Used as shared data for exchange API results between the steps.
    // Passed by reference to some step and may be updated in postProcess() of other.
    let localContext = [{
        id: context.user.userID,
        year: routeParams.year,
        month: routeParams.month,
        employerCode: 0
    }];
 
    const independent_steps = [
        new Step('me', 'get', '/me'),
        new Step('Docs', 'get', `/me/reports/${routeParams.year}/${routeParams.month}/docs`),
        new Step('Days Off', 'get', '/daysoff', 
            {
                year: routeParams.year,
                month: routeParams.month
            }),
        new Step('Manual Updates', 'get', '/me/reports/manual_updates',
            {
                year: routeParams.year,
                month: routeParams.month
            }),
        new Step('Report Data', 'get', `/me/reports/${routeParams.year}/${routeParams.month}`,
                null,
                (res) => {
                    localContext[0].employerCode = res.data.employerCode;
                })
    ]

    const dependent_steps = [
        new Step('Report Codes', 'get', '/me/report_codes',
                localContext[0]),
        new Step('Save Report', 'post', '/me/report/save',
            localContext[0])
    ]

    const [state, dispatch] = useReducer(reducer, [...independent_steps, ...dependent_steps]);

    const action_Update = (item) => {
        dispatch({
            type: 'UPDATE',
            payload: item
        })
    }
    
    const process_steps = async (steps) => {
        await Promise.all( steps.map( async(step) => {
            await step.process();
            action_Update(step);
            return step;
        }));        
    }

    const fetchData = async() => {

        // Firstly process independent steps.
        // At the finish of each step with postProccess() callback, 
        // the shared parameters are populated in order to be used when dependent steps are processed
        await process_steps(independent_steps);

        // At this point, all shared parameters are guaranteed to be ready
        await process_steps(dependent_steps);
    }

    useEffect( () => {
        fetchData();
    }, [])

    const onRefresh = () => {
        setLoading(true);
        fetchData();
        setLoading(false);
    }

    const activePanelChanged = (key) => {
        setActivePanelKeys(key)
    }

    const expandAllPanels = () => {
        
        expanded ?
            setActivePanelKeys([]) :
            setActivePanelKeys([...state.keys(), ]);  

        setExpanded(!expanded)
    }

    const genExtra = (status) => {

        const iconColor = status  == 200 ? 'rgb(24, 144, 255)' : 'rgb(255, 0, 0)';

        return <CheckCircleOutlined
                style={{
                    color: iconColor
                }} />
    }

    const sendOut = async() => {
        try {

            const _steps = state.map( step => {
                let _step = {...step, 
                    result: JSON.stringify(step.result),
                    callParams: JSON.stringify(step.params)
                }
                delete _step.params;
                return _step;
            })

            const incident = {
                steps: _steps
            };
            await API.post('/incidents', incident,
            {
                params: {
                    year: routeParams.year,
                    month: routeParams.month
                }
            })
        } catch(err) {
            throw err; // be catched in AppErrorBoundaries
        }
    }

    return (
        <Layout>
            <Layout.Header className='ant-layout-header-simple'>
                <Row>
                    <Col offset={9}>
                        <Title level={3}>{t('support_page')} {routeParams.year}/{routeParams.month}</Title>
                    </Col>
                </Row>
            </Layout.Header>
            <Layout.Content>
                <Row gutter={[0, 16]} type='flex'>
                    <Col span={22}>
                        <Button disabled={loading} onClick={onRefresh}>{t('refresh')}</Button>
                    </Col>
                    <Col span={2}>
                        <Button onClick={expandAllPanels}
                                style={{ float: 'right' }}>{
                           expanded? t('collapse') : t('expand')
                        }</Button>
                    </Col>
                </Row>
                <Row gutter={[0, 16]}>
                    <Col span={24}>
                        <Collapse 
                            onChange={activePanelChanged}
                            activeKey={activePanelKeys}>
                        {
                            state.map( (item, index) => {

                                return (
                                    <Panel header={item.name + ' (' + item.url + item.queryString + ')'} 
                                            key={index} 
                                            extra={genExtra(item.status)}>
                                        <div style={{
                                            backgroundColor: 'rgb(252, 246, 219)'
                                        }}>
                                            <pre>{JSON.stringify(item.result, null, 2)}</pre>
                                        </div>
                                    </Panel>
                                )
                            })
                        }
                        </Collapse>
                    </Col>
                </Row>
            </Layout.Content>
            <Layout.Footer>
                <Row>
                    <Col offset={10}>
                        <Button type='primary' onClick={sendOut}>Send to Support Pesonnel</Button>
                    </Col>
                </Row>
            </Layout.Footer>
        </Layout>
    )
}

export default SupportPage;