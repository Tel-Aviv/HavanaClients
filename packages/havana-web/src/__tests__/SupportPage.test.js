// 
// Support Page Test Suite
// 
// Test the apperiance on SupportPage Component based on varios API calls.
//
import 'babel-polyfill'

import React from 'react';
import { Route, Switch, HashRouter} from 'react-router-dom';
import renderer from 'react-test-renderer';
import SupportPage from '../components/SupportPage';

describe('Get data the API and display it', () => {

    test('Display SupportPage component', async() => {
        // await act( async () => mount(<HashRouter>
        //     <SupportPage />
        // </HashRouter>));
        // // const component = renderer.create(
        // //         <HashRouter>
        // //             <SupportPage />
        // //         </HashRouter>
        // // );

        // let tree = component.toJSON();
        // expect(tree).toMatchSnapshot();
    });

});