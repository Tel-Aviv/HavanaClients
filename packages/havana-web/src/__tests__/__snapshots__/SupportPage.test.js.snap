// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get data the API and display it Display SupportPage component 1`] = `
<section
  className="ant-layout"
>
  <header
    className="ant-layout-header-simple ant-layout-header"
  >
    <div
      className="ant-row"
      style={Object {}}
    >
      <div
        className="ant-col ant-col-offset-9"
        style={Object {}}
      >
        <h3
          aria-label={null}
          className="ant-typography"
          style={
            Object {
              "WebkitLineClamp": null,
            }
          }
        >
          Support Page for Report 
          /
        </h3>
      </div>
    </div>
  </header>
  <main
    className="ant-layout-content"
  >
    <div
      className="ant-row-flex"
      style={
        Object {
          "marginBottom": -8,
          "marginTop": -8,
        }
      }
    >
      <div
        className="ant-col ant-col-22"
        style={
          Object {
            "paddingBottom": 8,
            "paddingTop": 8,
          }
        }
      >
        <button
          className="ant-btn"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <span>
            Refresh
          </span>
        </button>
      </div>
      <div
        className="ant-col ant-col-2"
        style={
          Object {
            "paddingBottom": 8,
            "paddingTop": 8,
          }
        }
      >
        <button
          className="ant-btn"
          onClick={[Function]}
          style={
            Object {
              "float": "right",
            }
          }
          type="button"
        >
          <span>
            Expand All
          </span>
        </button>
      </div>
    </div>
    <div
      className="ant-row"
      style={
        Object {
          "marginBottom": -8,
          "marginTop": -8,
        }
      }
    >
      <div
        className="ant-col"
        style={
          Object {
            "paddingBottom": 8,
            "paddingTop": 8,
          }
        }
      >
        <div
          className="ant-collapse ant-collapse-icon-position-left"
          role={null}
        >
          <div
            className="ant-collapse-item"
          >
            <div
              aria-expanded="false"
              className="ant-collapse-header"
              onClick={[Function]}
              onKeyPress={[Function]}
              role="button"
              tabIndex={0}
            >
              <i
                aria-label="icon: right"
                className="anticon anticon-right ant-collapse-arrow"
              >
                <svg
                  aria-hidden="true"
                  className=""
                  data-icon="right"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M765.7 486.8L314.9 134.7A7.97 7.97 0 0 0 302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 0 0 0-50.4z"
                  />
                </svg>
              </i>
              me (/me)
              <div
                className="ant-collapse-extra"
              >
                <i
                  aria-label="icon: check-circle"
                  className="anticon anticon-check-circle"
                  style={
                    Object {
                      "color": "rgb(255, 0, 0)",
                    }
                  }
                >
                  <svg
                    aria-hidden="true"
                    className=""
                    data-icon="check-circle"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M699 353h-46.9c-10.2 0-19.9 4.9-25.9 13.3L469 584.3l-71.2-98.8c-6-8.3-15.6-13.3-25.9-13.3H325c-6.5 0-10.3 7.4-6.5 12.7l124.6 172.8a31.8 31.8 0 0 0 51.7 0l210.6-292c3.9-5.3.1-12.7-6.4-12.7z"
                    />
                    <path
                      d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                    />
                  </svg>
                </i>
              </div>
            </div>
          </div>
          <div
            className="ant-collapse-item"
          >
            <div
              aria-expanded="false"
              className="ant-collapse-header"
              onClick={[Function]}
              onKeyPress={[Function]}
              role="button"
              tabIndex={0}
            >
              <i
                aria-label="icon: right"
                className="anticon anticon-right ant-collapse-arrow"
              >
                <svg
                  aria-hidden="true"
                  className=""
                  data-icon="right"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M765.7 486.8L314.9 134.7A7.97 7.97 0 0 0 302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 0 0 0-50.4z"
                  />
                </svg>
              </i>
              Days Off (/daysoff)
              <div
                className="ant-collapse-extra"
              >
                <i
                  aria-label="icon: check-circle"
                  className="anticon anticon-check-circle"
                  style={
                    Object {
                      "color": "rgb(255, 0, 0)",
                    }
                  }
                >
                  <svg
                    aria-hidden="true"
                    className=""
                    data-icon="check-circle"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M699 353h-46.9c-10.2 0-19.9 4.9-25.9 13.3L469 584.3l-71.2-98.8c-6-8.3-15.6-13.3-25.9-13.3H325c-6.5 0-10.3 7.4-6.5 12.7l124.6 172.8a31.8 31.8 0 0 0 51.7 0l210.6-292c3.9-5.3.1-12.7-6.4-12.7z"
                    />
                    <path
                      d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                    />
                  </svg>
                </i>
              </div>
            </div>
          </div>
          <div
            className="ant-collapse-item"
          >
            <div
              aria-expanded="false"
              className="ant-collapse-header"
              onClick={[Function]}
              onKeyPress={[Function]}
              role="button"
              tabIndex={0}
            >
              <i
                aria-label="icon: right"
                className="anticon anticon-right ant-collapse-arrow"
              >
                <svg
                  aria-hidden="true"
                  className=""
                  data-icon="right"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M765.7 486.8L314.9 134.7A7.97 7.97 0 0 0 302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 0 0 0-50.4z"
                  />
                </svg>
              </i>
              Manual Updates (/me/reports/manual_updates)
              <div
                className="ant-collapse-extra"
              >
                <i
                  aria-label="icon: check-circle"
                  className="anticon anticon-check-circle"
                  style={
                    Object {
                      "color": "rgb(255, 0, 0)",
                    }
                  }
                >
                  <svg
                    aria-hidden="true"
                    className=""
                    data-icon="check-circle"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M699 353h-46.9c-10.2 0-19.9 4.9-25.9 13.3L469 584.3l-71.2-98.8c-6-8.3-15.6-13.3-25.9-13.3H325c-6.5 0-10.3 7.4-6.5 12.7l124.6 172.8a31.8 31.8 0 0 0 51.7 0l210.6-292c3.9-5.3.1-12.7-6.4-12.7z"
                    />
                    <path
                      d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                    />
                  </svg>
                </i>
              </div>
            </div>
          </div>
          <div
            className="ant-collapse-item"
          >
            <div
              aria-expanded="false"
              className="ant-collapse-header"
              onClick={[Function]}
              onKeyPress={[Function]}
              role="button"
              tabIndex={0}
            >
              <i
                aria-label="icon: right"
                className="anticon anticon-right ant-collapse-arrow"
              >
                <svg
                  aria-hidden="true"
                  className=""
                  data-icon="right"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M765.7 486.8L314.9 134.7A7.97 7.97 0 0 0 302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 0 0 0-50.4z"
                  />
                </svg>
              </i>
              Report Data (/me/reports/undefined/undefined)
              <div
                className="ant-collapse-extra"
              >
                <i
                  aria-label="icon: check-circle"
                  className="anticon anticon-check-circle"
                  style={
                    Object {
                      "color": "rgb(255, 0, 0)",
                    }
                  }
                >
                  <svg
                    aria-hidden="true"
                    className=""
                    data-icon="check-circle"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M699 353h-46.9c-10.2 0-19.9 4.9-25.9 13.3L469 584.3l-71.2-98.8c-6-8.3-15.6-13.3-25.9-13.3H325c-6.5 0-10.3 7.4-6.5 12.7l124.6 172.8a31.8 31.8 0 0 0 51.7 0l210.6-292c3.9-5.3.1-12.7-6.4-12.7z"
                    />
                    <path
                      d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                    />
                  </svg>
                </i>
              </div>
            </div>
          </div>
          <div
            className="ant-collapse-item"
          >
            <div
              aria-expanded="false"
              className="ant-collapse-header"
              onClick={[Function]}
              onKeyPress={[Function]}
              role="button"
              tabIndex={0}
            >
              <i
                aria-label="icon: right"
                className="anticon anticon-right ant-collapse-arrow"
              >
                <svg
                  aria-hidden="true"
                  className=""
                  data-icon="right"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M765.7 486.8L314.9 134.7A7.97 7.97 0 0 0 302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 0 0 0-50.4z"
                  />
                </svg>
              </i>
              Report Codes (/me/report_codes?id=308680768&employerCode=1&year=undefined&month=undefined)
              <div
                className="ant-collapse-extra"
              >
                <i
                  aria-label="icon: check-circle"
                  className="anticon anticon-check-circle"
                  style={
                    Object {
                      "color": "rgb(255, 0, 0)",
                    }
                  }
                >
                  <svg
                    aria-hidden="true"
                    className=""
                    data-icon="check-circle"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M699 353h-46.9c-10.2 0-19.9 4.9-25.9 13.3L469 584.3l-71.2-98.8c-6-8.3-15.6-13.3-25.9-13.3H325c-6.5 0-10.3 7.4-6.5 12.7l124.6 172.8a31.8 31.8 0 0 0 51.7 0l210.6-292c3.9-5.3.1-12.7-6.4-12.7z"
                    />
                    <path
                      d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                    />
                  </svg>
                </i>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </main>
  <footer
    className="ant-layout-footer"
  >
    <div
      className="ant-row"
      style={Object {}}
    >
      <div
        className="ant-col ant-col-offset-10"
        style={Object {}}
      >
        <button
          className="ant-btn ant-btn-primary"
          onClick={[Function]}
          type="button"
        >
          <span>
            Send to Support Pesonnel
          </span>
        </button>
      </div>
    </div>
  </footer>
</section>
`;
